src/sage/arith/misc.py

```diff
@@ -1842,76 +1865,29 @@ def gcd(a, b=None, **kwargs):
     """
     # Most common use case first:
     if b is not None:
-        try:
-            m = a.gcd
-        except (AttributeError, TypeError):
-            e = py_scalar_to_element(a)
-            if e is a:
-                raise TypeError("unable to call gcd with {!r}".format(a))
-            m = e.gcd
-
-        try:
-            return m(b, **kwargs)
-        except TypeError:
-            return m(py_scalar_to_element(b), **kwargs)
+        a = py_scalar_to_element(a)
+        b = py_scalar_to_element(b)
+        return a.gcd(b, **kwargs)
 
+    from sage.categories.objects import Objects
     from sage.structure.sequence import Sequence
     seq = Sequence(py_scalar_to_element(el) for el in a)
-    if seq.universe() is ZZ:
-        return GCD_list(seq)
-    else:
-        return __GCD_sequence(seq, **kwargs)
-
+    U = seq.universe()
 
-GCD = gcd
+    if len(seq) == 0:
+        return U(Integer(0))
 
+    if U is Objects():
+        raise ValueError("gcd arguments must share a common parent")
 
-def __GCD_sequence(v, **kwargs):
-    """
-    Internal function returning the gcd of the elements of a sequence
-
-    INPUT:
-
-
-    -  ``v`` - A sequence (possibly empty)
-
-
-    OUTPUT: The gcd of the elements of the sequence as an element of
-    the sequence's universe, or the integer 0 if the sequence is
-    empty.
-
-    EXAMPLES::
+    if U is ZZ:
+        return GCD_list(seq)
```

Comment:
This should be tested before `Objects()` as it is more likely to be called (if that test remains).